# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile sum.cpp
#include <iostream>
 
int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}
 
int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting sum.cpp


 Código de la suma de dos números en `C++`.

In [ ]:
!g++ sum.cpp -o sum
!./sum

^C


Comando para ejecutar el código de `C++` en Colab.

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
4

Overwriting output2.txt


Se crean los arichos `.txt` de la entrada y salida

In [ ]:
!./sum < input1.txt

12

In [ ]:
!./sum < input2.txt

4

Se prueban los archivos `.txt`

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt

12


A través del comando `grep` comparamos dos archivos y nos muestra si hay alguna diferencia entre alguno de los dos archivos.

In [ ]:
!./sum < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./sum < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

Creamos los test y comparamos con las salidas.

## Python

In [ ]:
%%writefile sum.py
# python3
 
 
def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit
 
if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


Código en Python de la suma de dos números.

In [ ]:
!python sum.py

Traceback (most recent call last):
  File "sum.py", line 8, in <module>
    a, b = map(int, input().split())
KeyboardInterrupt
^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

Stress testing para el programa de la suma de dos números.

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
 
int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();
 
    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }
 
    return max_product;
}
 
int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }
 
    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product1.cpp


Codigo en `C++` de la multiplicación de los dos números mayores de un arreglo.

In [ ]:
!g++ max_pairwise_product1.cpp -o max_pairwise_product1
!./max_pairwise_product1

^C


Ejecución del programa de el producto.

### Test 1: lagest input numbers 10000 10000 in the input

In [ ]:
%%writefile input1.txt
10
1 3 100000 9 100000 23

Overwriting input1.txt


In [ ]:
!./max_pairwise_product1 < input1.txt

1410065408


Debido a qué el código está manejando `int` cuando se multiplica $10.000$ por $10.000$ el resultado se sale de los límites de la variable.

In [ ]:
%%writefile max_pairwise_product2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product2.cpp


In [ ]:
!g++ max_pairwise_product2.cpp -o max_pairwise_product2
!./max_pairwise_product2 < input1.txt

10000000000


Se realiza el cambio de `int` a `long long` y el resultado está dentro de los límites de este.

### Test 2  generate a tests file with the maximun n 100000 numbers and measure time

In [ ]:
%%writefile nmax.cpp
#include <iostream>
#include <vector>
#include <algorithm>


    int main(){
        int n=100000;
        std::cout<< n <<"\n";
        int number;

      for(int i=0;i<n;++i){
        number = 1 + rand() % (100000 - 1);
        std::cout<<number<< " ";
      }
        
        return 0;
    }
		

Overwriting nmax.cpp


In [ ]:
!g++ nmax.cpp -o nmax
!./nmax > inputT2.txt

Se realiza un archivo de entradas con $100.000$ números al azar

### Efficient solution 

In [ ]:
%%writefile max_pairwise_productFast.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int main() {

	
    int n;
    std::cin >> n;
    long long max_numb = 0LL;
    long long sec_max_numb = 0LL;
    long long temporal=0LL;
    long long max_product=0LL;


    for(int i=0;i<n; ++i){
        
    	std::cin>> temporal;

      if(temporal>=max_numb){
          sec_max_numb=max_numb;
          max_numb=temporal;
      }
      else if(temporal>sec_max_numb){
          sec_max_numb=temporal;
      }
	  }


    max_product=((long long)max_numb) * ((long long) sec_max_numb);

  	std::cout<<max_product<<"\n";
    
}

Overwriting max_pairwise_productFast.cpp


  Codigo optimizado de el programa de la multiplicacion de los numeros mayores de una arreglo, Cambiando el tiempo de ejecucion de $O(n^2)$ a $O(n)$

In [ ]:
%%time
!g++ max_pairwise_productFast.cpp -o max_pairwise_productFast
!./max_pairwise_productFast < input1.txt

10000000000
CPU times: user 12.4 ms, sys: 12.1 ms, total: 24.5 ms
Wall time: 531 ms


In [ ]:
%%time
!g++ max_pairwise_productFast.cpp -o max_pairwise_productFast
!./max_pairwise_productFast < inputT2.txt

9999500004
CPU times: user 18.8 ms, sys: 12.2 ms, total: 31 ms
Wall time: 542 ms


Comparar el tiempo con la solución anterior 


In [ ]:
%%time
!g++ max_pairwise_product2.cpp -o max_pairwise_product2
!./max_pairwise_product2 < inputT2.txt

^C
CPU times: user 58.1 ms, sys: 22.1 ms, total: 80.2 ms
Wall time: 8.14 s




*   Solución eficiente: 543 ms
*   Solución anterior: 32.4 s



### Stress tests 

In [ ]:
!./max_pairwise_product2 < inputT2.txt > stresstest1.txt
!./max_pairwise_productFast < inputT2.txt > stresstest2.txt
!grep -xvFf stresstest1.txt stresstest2.txt

## Python

In [ ]:
%%writefile max_pairwise_product1.py 
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product


if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))

Overwriting max_pairwise_product1.py


In [ ]:
!python max_pairwise_product1.py

Traceback (most recent call last):
  File "max_pairwise_product1.py", line 13, in <module>
    input_n = int(input())
KeyboardInterrupt
^C


Codigo en `Python` de la multiplicación de los dos números mayores de un arreglo.

### Test 1 


In [ ]:
%%writefile input1.txt
6
1 3 100000 9 100000 23

Overwriting input1.txt


In [ ]:
!python max_pairwise_product1.py < input1.txt

10000000000


Prueba de codigo hecho en `Python`.

### Efficient solution

In [ ]:
%%writefile max_pairwise_product2.py 

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    max_numb=0
    second_max_numb=0
    max_product=0
    for temporal in range(input_n):
      if (input_numbers[temporal]>=max_numb):
        second_max_numb=max_numb
        max_numb=input_numbers[temporal]
      elif (input_numbers[temporal]>second_max_numb):
        second_max_numb=input_numbers[temporal]

    max_product= max_numb * second_max_numb
    print(max_product)

Writing max_pairwise_product2.py


In [ ]:
!python max_pairwise_product2.py

Traceback (most recent call last):
  File "max_pairwise_product2.py", line 3, in <module>
    input_n = int(input())
KeyboardInterrupt
^C


In [ ]:
!python max_pairwise_product2.py < input1.txt

10000000000


  Codigo optimizado de el programa de la multiplicacion de los numeros mayores de una arreglo, Cambiando el tiempo de ejecucion de $O(n^2)$ a $O(n)$

### Test 2

In [ ]:
%%time
!python max_pairwise_product1.py < inputT2.txt

9999500004
CPU times: user 7.85 s, sys: 1.06 s, total: 8.91 s
Wall time: 21min 5s


In [ ]:
%%time
!python max_pairwise_product2.py < inputT2.txt

9999500004
CPU times: user 9.35 ms, sys: 4.89 ms, total: 14.2 ms
Wall time: 215 ms


*   Solución eficiente: 215 ms
*   Solución anterior: 21min 5s


### Stress Test

In [ ]:
!python max_pairwise_product2.py < inputT2.txt> pstresstest1.txt
!grep -xvFf pstresstest1.txt stresstest2.txt